concurrency/PendingReconstructionBlocks.java

```diff
@@ -0,0 +1,306 @@
+   * system started reconstructing the most recent copy of this
+   * block. It also records the list of Datanodes where the
+   * reconstruction requests are in progress.
+   */
+  static class PendingBlockInfo {
+    private long timeStamp;
+    private final List<DatanodeDescriptor> targets;
+
+    PendingBlockInfo(DatanodeDescriptor[] targets) {
+      this.timeStamp = monotonicNow();
+      this.targets = targets == null ? new ArrayList<DatanodeDescriptor>()
+          : new ArrayList<>(Arrays.asList(targets));
+    }
+
+    long getTimeStamp() {
+      return timeStamp;
+    }
+
+    void setTimeStamp() {
+      timeStamp = monotonicNow();
+    }
+
+    void incrementReplicas(DatanodeDescriptor... newTargets) {
+      if (newTargets != null) {
+        for (DatanodeDescriptor newTarget : newTargets) {
+          if (!targets.contains(newTarget)) {
+            targets.add(newTarget);
+          }
+        }
+      }
+    }
+
+    void decrementReplicas(DatanodeDescriptor dn) {
+      targets.remove(dn);
+    }
+
+    int getNumReplicas() {
+      return targets.size();
+    }
+  }
+
+  /*
+   * A periodic thread that scans for blocks that never finished
+   * their reconstruction request.
+   */
+  class PendingReconstructionMonitor implements Runnable {
+    @Override
+    public void run() {
+      while (fsRunning) {
+        long period = Math.min(DEFAULT_RECHECK_INTERVAL, timeout);
+        try {
+          pendingReconstructionCheck();
+          Thread.sleep(period);
+        } catch (InterruptedException ie) {
+          LOG.debug("PendingReconstructionMonitor thread is interrupted.", ie);
+        }
+      }
+    }
+
+    /**
+     * Iterate through all items and detect timed-out items
+     */
+    void pendingReconstructionCheck() {
+      synchronized (pendingReconstructions) {
+        Iterator<Map.Entry<BlockInfo, PendingBlockInfo>> iter =
+            pendingReconstructions.entrySet().iterator();
+        long now = monotonicNow();
+        LOG.debug("PendingReconstructionMonitor checking Q");
+        while (iter.hasNext()) {
+          Map.Entry<BlockInfo, PendingBlockInfo> entry = iter.next();
+          PendingBlockInfo pendingBlock = entry.getValue();
+          if (now > pendingBlock.getTimeStamp() + timeout) {
+            BlockInfo block = entry.getKey();
+            synchronized (timedOutItems) {
+              timedOutItems.add(block);
+            }
+            LOG.warn("PendingReconstructionMonitor timed out " + block);
+            NameNode.getNameNodeMetrics().incTimeoutReReplications();
+            iter.remove();
+          }
+        }
+      }
+    }
+  }
+
+  /*
+   * Shuts down the pending reconstruction monitor thread.
+   * Waits for the thread to exit.
+   */
+  void stop() {
+    fsRunning = false;
+    if(timerThread == null) return;
+    timerThread.interrupt();
+    try {
+      timerThread.join(3000);
+    } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)